internal/app/upstream/client.go

```diff
@@ -55,11 +74,179 @@ type Response struct {
+	}
+
+	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
+	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
+	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
+	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
+
+	go shutDown(ctx, conn)
+
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+		logger:      logger,
+	}, nil
+}
+
+func (m *client) OpenStream(request v2.DiscoveryRequest) (<-chan *v2.DiscoveryResponse, func(), error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	var stream grpc.ClientStream
+	var err error
+	switch request.GetTypeUrl() {
+	case ListenerTypeURL:
+		stream, err = m.ldsClient.StreamListeners(ctx)
+	case ClusterTypeURL:
+		stream, err = m.cdsClient.StreamClusters(ctx)
+	case RouteTypeURL:
+		stream, err = m.rdsClient.StreamRoutes(ctx)
+	case EndpointTypeURL:
+		stream, err = m.edsClient.StreamEndpoints(ctx)
+	default:
+		cancel()
+		m.logger.Error(ctx, "Unsupported Type Url %s", request.GetTypeUrl())
+		return nil, nil, &UnsupportedResourceError{TypeURL: request.GetTypeUrl()}
+	}
+
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+
+	signal := make(chan *version, 1)
+	// The xds protocol https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack
+	// specifies that the first request be empty nonce and empty version.
+	// The origin server will respond with the latest version.
+	signal <- &version{nonce: "", version: ""}
+
+	response := make(chan *v2.DiscoveryResponse)
+
+	go send(ctx, m.logger, cancel, &request, stream, signal, m.callOptions)
+	go recv(ctx, cancel, m.logger, response, stream, signal)
+
+	// We use context cancellation over using a separate channel for signalling stream shutdown.
+	// The reason is cancelling a context tied with the stream is straightforward to signal closure.
+	// Also, the shutdown function could potentially be called more than once by a caller.
+	// Closing channels is not idempotent while cancelling context is idempotent.
+	return response, func() { cancel() }, nil
+}
+
+// It is safe to assume send goroutine will not leak as long as these conditions are true:
+// - SendMsg is performed with timeout.
+// - send is a receiver for signal and exits when signal is closed by the owner.
+// - send also exits on context cancellations.
+func send(
+	ctx context.Context,
+	logger log.Logger,
+	cancelFunc context.CancelFunc,
+	request *v2.DiscoveryRequest,
+	stream grpc.ClientStream,
+	signal chan *version,
+	callOptions CallOptions) {
+	for {
+		select {
+		case sig, ok := <-signal:
+			if !ok {
+				return
+			}
+			request.ResponseNonce = sig.nonce
+			request.VersionInfo = sig.version
+			err := doWithTimeout(ctx, func() error {
+				return stream.SendMsg(request)
+			}, callOptions.Timeout)
+			if err != nil {
+				select {
+				case <-ctx.Done():
+					// Context was cancelled, hence this is not an erroneous scenario.
```

Comment:
Just so I understand, this only can happen during a timeout?

Reply:
A timeout from `doWithTimeout` will will cause an error [here](https://github.com/envoyproxy/xds-relay/blob/0ac6a265387b46e81d5f10805ef06e1f8398b82d/internal/app/upstream/client.go#L242).
In the test cases i figured there's racecondition and shutdown could be called when timeout happens. In that case since shutdown causes the parent context cancellation, its prudent to check on it and return. Since the shutdown is already called, there is no point in reporting an error.

Reply:
In other words throughout the implementation `ctx.Done()` happens only when `shutdown()` has been called. When goroutines get unblocked and find shutdown has been called, they should exit instead of continuing. I will put the intent in a comment.

Reply:
added some more comments [here](https://github.com/envoyproxy/xds-relay/blob/f5a366b4816689977fc082da1e07a2f71377cf99/internal/app/upstream/client.go#L48-L50).

Reply:
I think this is the remaining one. @lita lmk what you feel about the explanation.

Reply:
I would add a comment around that the parent context was canceled, probably due to shutdown.

It would also be good to abstract this in a method as these nested selects are hard to read.

```
// Handles the scenario when sendMsg has an error
func handleSendErr(ctx context.Context, err error.Error, cancelFunc func) {
	defer cancelFunc()

	select {
	// If the context is done, then we are probably shutting down. Don't log the error and continue with shutdown.
	case <- ctx.Done():
		return 
	}

	logger.Error(ctx, "Error in SendMsg: %s", err.Error())

}
```

Reply:
Looks like you have the same logic in recv as well which can leverage this code:

https://github.com/envoyproxy/xds-relay/pull/46/files#diff-c30e80dfa8b61ea3ccdc1ae6d3716890R198-R204

Reply:
pulled the common piece into a function.